Setup assistant automating generation of controllers configs

MohmadAyman opened the pull request:
### Description

This PR automates the generation of the ros_controller config files through the Setup Assistant, this is a part of upgrading the setup assistant.  [Issue 894](https://github.com/ros-planning/moveit/issues/894)

This [ros_controllers.yaml](https://gist.github.com/MohmadAyman/0e02336956cdec3b270d0e49f386b6b5) was produced using this PR foe the PR2 Robot. 

I'll follow this PR with another one that exposes the options through a GUI, and add documentation for it.

### Checklist
- [ ] **Required**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Include a screenshot if changing a GUI
- [ ] Optional: Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Optional: Decide if this should be cherry-picked to other current ROS branches (Indigo, Jade, Kinetic)

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

davetcoleman commented on moveit_setup_assistant/src/widgets/configuration_files_widget.cpp:
```diff
@@ -284,12 +284,11 @@ bool ConfigurationFilesWidget::loadGenFiles()
-  // fake_controllers.yaml --------------------------------------------------------------------------------------
-  file.file_name_ = "fake_controllers.yaml";
+  // controllers.yaml --------------------------------------------------------------------------------------
+  file.file_name_ = "controllers.yaml";
```
ros_controller.yaml

MohmadAyman commented:
https://gist.github.com/MohmadAyman/76abce5525170ec6e0b119787c44490f

davetcoleman commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -251,7 +251,7 @@ class MoveItConfigData
   bool outputOMPLPlanningYAML(const std::string& file_path);
   bool outputKinematicsYAML(const std::string& file_path);
   bool outputJointLimitsYAML(const std::string& file_path);
-  bool outputFakeControllersYAML(const std::string& file_path);
```
lets keep the fake controllers also

davetcoleman commented on moveit_setup_assistant/templates/moveit_config_pkg_template/launch/controllers.launch:
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0"?>
```
ros_controllers.launch

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+{
+  // Cache the joints' names.
+  std::vector<std::vector<std::string>> groups;
+  std::vector<std::string> groupJoints;
```
please use underscore for var names per ros style guidelines

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
+{
+  // Cache the joints' names.
+  std::vector<std::vector<std::string>> groups;
```
use more specific name than just groups

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+  std::vector<std::vector<std::string>> groups;
+  std::vector<std::string> groupJoints;
+
+  // Loop through groups
```
expand comment: why loop through?

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+      if (joint->isPassive() || joint->getMimic() != NULL || joint->getType() == robot_model::JointModel::FIXED)
+        continue;
+      else
+      {
```
don't mix using brackets without using brackets... all or none

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+    // Get list of associated joints
+    const robot_model::JointModelGroup* joint_model_group = getRobotModel()->getJointModelGroup(group_it->name_);
+    const std::vector<const robot_model::JointModel*>& joint_models = joint_model_group->getActiveJointModels();
+    // Iterate through the joints
```
explain why you are iterating through the joints

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+  emitter << YAML::Value << YAML::BeginSeq;
+
+  // Loop through groups
+  for (int i = 0; i < srdf_->groups_.size(); ++i)
```
std::size_t

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+    emitter << YAML::Value << YAML::BeginSeq;
+
+    // Iterate through the joints
+    for (int j = 0; j < groups[i].size(); j++)
```
std::size_t

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+  emitter << YAML::Newline << YAML::Comment("Creates the /joint_states topic necessary in ROS");
+  emitter << YAML::EndMap;
+
+  emitter << YAML::BeginMap;
```
what if in order to track the begin and end of maps, brackets were used to indent these sections? e.g.

```
emitter << YAML::BeginMap;
{
    emitter << YAML::Key << "joint_state_controller";
    emitter << YAML::Value << YAML::BeginMap;
    ...
}
emitter << YAML::EndMap;
```

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+  output_stream << emitter.c_str();
+  output_stream.close();
+
+  return true;  // file created successfully
```
this function is very long and readability is low. can it be separated out into a few smaller functions for each section of the yaml?

davetcoleman replied:
Sure, no problem.

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -554,6 +554,213 @@ bool MoveItConfigData::outputFakeControllersYAML(const std::string& file_path)
+  std::ofstream output_stream(file_path.c_str(), std::ios_base::trunc);
+  if (!output_stream.good())
+  {
+    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
```
use ROS_ERROR_STREAM_NAMED

MohmadAyman replied:
Why do you recommend using the Named version here?
All the functions here that cant output a file uses this
`ROS_ERROR_STREAM("Unable to open file for writing " << file_path);`

davetcoleman replied:
See the style guidelines under "ROS":
http://moveit.ros.org/documentation/contributing/code/

It makes filtering all the command line output much easier and cleaner. All of the MSA should be updated to use NAMED, but that belongs in a separate simple PR that doesn't clutter this one.

MohmadAyman replied:
Okay I understand, Thanks.

davetcoleman commented on moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch:
```diff
@@ -0,0 +1,17 @@
+  <!-- Load joint controller configurations from YAML file to parameter server -->
+  <rosparam file="$(find [GENERATED_PACKAGE_NAME])/config/controllers.yaml" command="load"/>
+
+  <!-- load the controllers -->
```
be consistent with capitalization in comments. i personally prefer capitalizing

davetcoleman commented:
@mcevoyandy please review

MohmadAyman commented:
Sorry didn't see the comments till now, I'll have a look at them later today.
For now I got the output configuration to work with '[moveit_sim_controller](https://github.com/davetcoleman/moveit_sim_controller)'.

MohmadAyman commented:
Addressed the comments, please have a look. 
this is a [ros_conttollers.yaml](https://gist.github.com/MohmadAyman/0e02336956cdec3b270d0e49f386b6b5) produced for the ur5 and tested with moveit_sim_controller
@davetcoleman  @mcevoyandy

rhaschke commented on moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch:
```diff
@@ -0,0 +1,17 @@
+
+  <!-- Load the controllers -->
+  <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
+    output="screen" ns="/[GENERATED_PACKAGE_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
```
@willcbaker Is using absolute namespaces compatible to 4598f6602ded91d91fccabe7a8bcc490a528fc54?

willcbaker replied:
I don't believe so, I would advise against absolute namespacing.
Unable to test or review until next week, seems changing this way with generated package name will require pushing down into package name namespace?

MohmadAyman replied:
So you recommend to remove the `ns` argument from here and from the `robot_state_publisher` node? right?

rhaschke replied:
Yes, I think so.

MohmadAyman replied:
This `ros_controllers.launch` is intended to be used with `gazebo`, and according to [`ros_control` tutorials](http://gazebosim.org/tutorials/?tut=ros_control#Addthegazebo_ros_controlplugin), this is how `gazebo_ros_control` plugin is defined in the `urdf`
```
<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
    <robotNamespace>/MYROBOT</robotNamespace>
  </plugin>
</gazebo>
```
So if we were to launch the controllers with not the `robot_name` as a `namespace` , and leave the plugin element `<robotNamespace>`  blank, how would we be able to use simulate multiple robot for example?
Is there a solution for a situation like that?

rhaschke commented on moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch:
```diff
@@ -0,0 +1,17 @@
+  <!-- Convert joint states to TF transforms for rviz, etc -->
+  <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher"
+    respawn="false" output="screen">
+    <remap from="/joint_states" to="/[GENERATED_PACKAGE_NAME]/joint_states" />
```
Same here.

rhaschke commented on moveit_setup_assistant/include/moveit/setup_assistant/tools/moveit_config_data.h:
```diff
@@ -254,6 +254,7 @@ class MoveItConfigData
   bool outputKinematicsYAML(const std::string& file_path);
   bool outputJointLimitsYAML(const std::string& file_path);
   bool outputFakeControllersYAML(const std::string& file_path);
+  bool outputControllersYAML(const std::string& file_path);
```
outputROSControllersYAML?

davetcoleman commented on moveit_setup_assistant/src/tools/moveit_config_data.cpp:
```diff
@@ -732,7 +732,7 @@ bool MoveItConfigData::outputControllersYAML(const std::string& file_path)
   if (!output_stream.good())
   {
-    ROS_ERROR_STREAM("Unable to open file for writing " << file_path);
+    ROS_ERROR_STREAM_NAMED("ros_controllers_config", "Unable to open file for writing " << file_path);
```
how about namespace = "setup_assistant"

rhaschke replied:
-1 The package name (which is moveit_setup_assistant) is already part of the logger name. No need for an additional name at all!

davetcoleman replied:
Good point about the duplicate namespace, I should have recommended the file name "moveit_config_data" as that's the standard we usually follow. I really do like using _NAMED though. I've just attempted to document this standard better:
https://github.com/ros-planning/moveit.ros.org/pull/166

MohmadAyman replied:
So do we get back to `ROS_ERROR_STREAM`?

rhaschke replied:
That's my suggestion. @davetcoleman?

davetcoleman replied:
I prefer the consistency of always using NAMED as I've just attempted to document in https://github.com/ros-planning/moveit.ros.org/pull/166. Perhaps a Github issue should be opened to specifically debate this. Either way, having the namespace certainly does not hurt anything so I would like to just merge this PR.

rhaschke replied:
@davetcoleman I don't agree. If you just have a redundant logger name like xxx.xxx, there is no reason to insist in `NAMED`. When we recently reworked the logging in some other MoveIt packages, we also dropped the extra NAME in such cases.

rhaschke commented on moveit_setup_assistant/templates/moveit_config_pkg_template/launch/ros_controllers.launch:
```diff
@@ -6,12 +6,12 @@
   <!-- Load the controllers -->
   <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false"
-    output="screen" ns="/[GENERATED_PACKAGE_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
+    output="screen" ns="[ROBOT_NAME]" args="spawn joint_state_controller position_trajectory_controller"/>
```
Looks like the ns prefix is very specific to your application. Why do you prefer `ROBOT_NAME` over `GENERATED_PACKAGE_NAME` now?

MohmadAyman replied:
After looking at [gazebo_ros_demos/rrbot_control.launch](https://github.com/ros-simulation/gazebo_ros_demos/blob/kinetic-devel/rrbot_control/launch/rrbot_control.launch) , I've found we use the robot name not the package name for the namespace.
Which is also done in the [ros_control tutorial here](http://gazebosim.org/tutorials/?tut=ros_control].

rhaschke commented:
@MohmadAyman Please do not merge the devel branch into pull-request branches - never ever.
Instead you should rebase your PR to the latest devel branch. 
Otherwise the commit history is screwed up and we have a really hard time to follow in the git tree what happened. Also, we cannot easily squash-merge the PR.

I manually merged this now in 89f55aaa7721351ce59cb1c0c09bd2b4141988c2.
